cpp/src/endpoint.cpp

```diff
@@ -191,6 +192,27 @@ void Endpoint::removeInflightRequest(const Request* const request)
 
 size_t Endpoint::cancelInflightRequests() { return _inflightRequests->cancelAll(); }
 
+std::shared_ptr<Request> Endpoint::amSend(void* buffer,
+                                          size_t length,
+                                          ucs_memory_type_t memoryType,
+                                          const bool enablePythonFuture,
+                                          RequestCallbackUserFunction callbackFunction,
+                                          RequestCallbackUserData callbackData)
+{
+  auto endpoint = std::dynamic_pointer_cast<Endpoint>(shared_from_this());
```

Comment:
Why do we need to cast `this` (which is an `Endpoint`?) to `Endpoint` (or is it that it might be a subclass?)